src/providers/aws.rs

```diff
@@ -727,41 +735,69 @@ impl RegionLauncher {
     /// Make spot instance requests, wait for the instances, and then call the
     /// instance setup functions.
     #[instrument(level = "debug", skip(self, max_wait))]
-    pub async fn launch<M>(
-        &mut self,
-        mode: LaunchMode,
+    pub async fn launch<'l, M>(
+        &'l mut self,
+        mut mode: LaunchMode,
         mut max_wait: Option<time::Duration>,
         machines: M,
     ) -> Result<(), Report>
     where
         M: IntoIterator<Item = (String, Setup)> + std::fmt::Debug,
     {
-        match mode {
-            LaunchMode::DefinedDuration {
-                hours: max_instance_duration_hours,
-            } => {
-                self.make_spot_instance_requests(
-                    max_instance_duration_hours * 60, // 60 mins/hr
-                    machines,
-                )
-                .await
-                .wrap_err("failed to make spot instance requests")?;
+        let machines: Vec<_> = machines.into_iter().collect();
 
-                let start = time::Instant::now();
-                self.wait_for_spot_instance_requests(max_wait)
-                    .await
-                    .wrap_err("failed while waiting for spot instances fulfilment")?;
-                if let Some(ref mut d) = max_wait {
-                    *d -= time::Instant::now().duration_since(start);
+        loop {
+            let machines = machines.clone();
+            match mode {
+                LaunchMode::TrySpot {
+                    hours: max_instance_duration_hours,
                 }
-            }
-            LaunchMode::OnDemand => {
-                self.make_on_demand_requests(machines)
+                | LaunchMode::DefinedDuration {
+                    hours: max_instance_duration_hours,
+                } => {
+                    self.make_spot_instance_requests(
+                        max_instance_duration_hours * 60, // 60 mins/hr
+                        machines,
+                    )
                     .await
-                    .wrap_err("failed to start on demand instances")?;
+                    .wrap_err("failed to make spot instance requests")?;
+
+                    let start = time::Instant::now();
+                    if let Err(e) = self
+                        .wait_for_spot_instance_requests(max_wait)
+                        .await
+                        .wrap_err(eyre!(
+                            "failed while waiting for spot instances fulfilment in {}",
+                            self.region.name()
+                        ))
+                    {
+                        if let LaunchMode::TrySpot { .. } = mode {
+                            mode = LaunchMode::OnDemand;
+                            tracing::debug!(err = ?e, "re-trying with OnDemand instace");
+                            continue;
```

Comment:
This doesn't seem quite right. What if some spot instances started but not others? I don't think any code will actually clean those up with the code as it is now?

Reply:
Ahh, in my head everything still gets cleaned up on drop.

Reply:
actually, `wait_for_spot_instance_requests` will handle this via `cancel_spot_instance_requets` I think.

Reply:
Ah, yes, I think you're right!